src/ui/speechrecognition/speechrecognizer.js

```diff
@@ -0,0 +1,146 @@
+     * @type {string}
+     */
+    this._latestResult = '';
+
+    /**
+     * The timeout ID of the silence threshold timeout
+     * @type {number|null}
+     */
+    this._silenceThresholdTimeout = null;
+
+    /**
+     * The amount of silence after the last detected word before triggering a search.
+     * This is primarily used for Safari since Edge and Chrome typically detect the end of
+     * voice search phrase very quickly.
+     */
+    this._silenceThresholdToSearch = 1000;
+
+    /**
+     * Indicates that speech recognition is currently active.
+     * This is used to prevent the speech recognizer from starting while it is already active.
+     * @type {boolean}
+     */
+    this._recognitionActive = false;
+
+    this._init();
+  }
+
+  _init () {
+    this._speechRecognition.addEventListener('start', () => {
+      this._recognitionActive = true;
+    });
+    this._speechRecognition.addEventListener('end', () => {
+      this._recognitionActive = false;
+      this._onCompleteHandler && this._onCompleteHandler(this._latestResult);
+    });
+    this._speechRecognition.addEventListener('result', event => {
+      this._handleResultEvent(event);
+    });
+    this._speechRecognition.addEventListener('error', event => {
+      this._handleErrorEvent(event);
+    });
+  }
+
+  /**
+   * Handles a search recognition result event
+   * @param {SpeechRecognitionEvent} event
+   */
+  _handleResultEvent (event) {
+    /**
+     * Set a timeout to stop the speech recognition if no words are detected for the configured amount
+     * of time. Every time a result comes in, reset that timeout.
+     */
+    clearTimeout(this._silenceThresholdTimeout);
+    this._silenceThresholdTimeout = setTimeout(() => {
+      this.stop();
+    }, this._silenceThresholdToSearch);
+
+    const result = event.results[0][0].transcript;
+    const isFinalResult = event.results[0].isFinal;
+
+    if (result !== this._latestResult) {
+      this._onResultHandler && this._onResultHandler(result);
+      this._latestResult = result;
+    }
+    if (isFinalResult) {
+      this.stop();
+    }
+  }
+
+  /**
+   * Handles a speech recognition error
+   * @param {SpeechRecognitionError} event
+   */
+  _handleErrorEvent (event) {
+    if (event.error === 'service-not-allowed') {
+      window.alert(
+        TranslationFlagger.flag({ phrase: 'Speech Recognition is not available.' }) + '\n' +
+        TranslationFlagger.flag({ phrase: 'Siri must be enabled for support on Safari.' })
+      );
+    } else {
+      console.warn(event);
+    }
+    this._onError && this._onError();
+  }
+
+  /**
+   * Starts the speech recognizer if it has not already started
+   */
+  start () {
+    try {
+      !this._recognitionActive && this._speechRecognition.start();
+    } catch (err) {
+      console.warn(err);
+    }
+  }
+
+  /**
+   * Stops the speech recognizer if it is active
+   */
+  stop () {
+    this._recognitionActive && this._speechRecognition.stop();
```

Comment:
Should this._recognitionActive be set to false?

Reply:
oh wait nvm, I guess this._speechRecognition.stop() trigger an 'end' event that would set this to false.

Reply:
Correct, I found it simplest to just have the 'start' and 'end' events toggle `this._recognitionActive'